drivers/net/nsos_adapt.c

```diff
@@ -0,0 +1,673 @@
+		*type = SOCK_RAW;
+		break;
+	default:
+		nsi_print_warning("%s: socket type %d not supported\n", __func__, type_mid);
+		return -NSOS_MID_ESOCKTNOSUPPORT;
+	}
+
+	return 0;
+}
+
+static int socket_type_to_nsos_mid(int type, int *type_mid)
+{
+	switch (type) {
+	case SOCK_STREAM:
+		*type_mid = NSOS_MID_SOCK_STREAM;
+		break;
+	case SOCK_DGRAM:
+		*type_mid = NSOS_MID_SOCK_DGRAM;
+		break;
+	case SOCK_RAW:
+		*type_mid = NSOS_MID_SOCK_RAW;
+		break;
+	default:
+		nsi_print_warning("%s: socket type %d not supported\n", __func__, type);
+		return -NSOS_MID_ESOCKTNOSUPPORT;
+	}
+
+	return 0;
+}
+
+static int socket_flags_from_nsos_mid(int flags_mid)
+{
+	int flags = 0;
+
+	nsos_socket_flag_convert(&flags_mid, NSOS_MID_MSG_PEEK,
+				 &flags, MSG_PEEK);
+	nsos_socket_flag_convert(&flags_mid, NSOS_MID_MSG_TRUNC,
+				 &flags, MSG_TRUNC);
+	nsos_socket_flag_convert(&flags_mid, NSOS_MID_MSG_DONTWAIT,
+				 &flags, MSG_DONTWAIT);
+	nsos_socket_flag_convert(&flags_mid, NSOS_MID_MSG_WAITALL,
+				 &flags, MSG_WAITALL);
+
+	if (flags_mid != 0) {
+		return -NSOS_MID_EINVAL;
+	}
+
+	return flags;
+}
+
+int nsos_adapt_socket(int family_mid, int type_mid, int proto_mid)
+{
+	int family;
+	int type;
+	int proto;
+	int ret;
+
+	ret = socket_family_from_nsos_mid(family_mid, &family);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = socket_type_from_nsos_mid(type_mid, &type);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = socket_proto_from_nsos_mid(proto_mid, &proto);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = socket(family, type, proto);
+	if (ret < 0) {
+		return -errno_to_nsos_mid(errno);
+	}
+
+	return ret;
+}
+
+static int sockaddr_from_nsos_mid(struct sockaddr **addr, socklen_t *addrlen,
+				  const struct nsos_mid_sockaddr *addr_mid, size_t addrlen_mid)
+{
+	if (!addr_mid) {
+		*addr = NULL;
+		*addrlen = 0;
+
+		return 0;
+	}
+
+	switch (addr_mid->sa_family) {
+	case NSOS_MID_AF_INET: {
+		const struct nsos_mid_sockaddr_in *addr_in_mid =
+			(const struct nsos_mid_sockaddr_in *)addr_mid;
+		struct sockaddr_in *addr_in = (struct sockaddr_in *)*addr;
+
+		addr_in->sin_family = AF_INET;
+		addr_in->sin_port = addr_in_mid->sin_port;
+		addr_in->sin_addr.s_addr = addr_in_mid->sin_addr;
+
+		*addrlen = sizeof(*addr_in);
+
+		return 0;
+	}
+	case NSOS_MID_AF_INET6: {
+		const struct nsos_mid_sockaddr_in6 *addr_in_mid =
+			(const struct nsos_mid_sockaddr_in6 *)addr_mid;
+		struct sockaddr_in6 *addr_in = (struct sockaddr_in6 *)*addr;
+
+		addr_in->sin6_family = AF_INET6;
+		addr_in->sin6_port = addr_in_mid->sin6_port;
+		addr_in->sin6_flowinfo = 0;
+		memcpy(addr_in->sin6_addr.s6_addr, addr_in_mid->sin6_addr,
+		       sizeof(addr_in->sin6_addr.s6_addr));
+		addr_in->sin6_scope_id = addr_in_mid->sin6_scope_id;
+
+		*addrlen = sizeof(*addr_in);
+
+		return 0;
+	}
+	}
+
+	return -NSOS_MID_EINVAL;
+}
+
+static int sockaddr_to_nsos_mid(const struct sockaddr *addr, socklen_t addrlen,
+				struct nsos_mid_sockaddr *addr_mid, size_t *addrlen_mid)
+{
+	if (!addr) {
```

Comment:
Would a check for `addrlen == 0` be necessary?
(similar comment in `sockaddr_from_nsos_mid()`)
Running the new sample under valgrind gives:
```
==3113887==    at 0x80689CB: sockaddr_to_nsos_mid (nsos_adapt.c:303)
==3113887==    by 0x8068E8C: nsos_adapt_recvfrom (nsos_adapt.c:451)
==3113887==    by 0x80617FB: nsos_recvfrom_with_poll (nsos_sockets.c:691)
==3113887==    by 0x80617FB: nsos_recvfrom (nsos_sockets.c:717)
==3113887==    by 0x804E14F: z_impl_zsock_recvfrom (sockets.c:1773)
==3113887==    by 0x8049756: zsock_recvfrom (socket.h:313)
==3113887==    by 0x8049756: zsock_recv (socket.h:547)
==3113887==    by 0x8049756: recv (socket.h:890)
==3113887==    by 0x8049756: main (http_get.c:115)
```

Reply:
Indeed, that was necessary. Stumbled upon that after checking error codes from `sockaddr_to_nsos_mid()`. So... added check for `addrlen == 0`, as suggested.